toplevel/topdirs.ml

```diff
@@ -549,6 +549,22 @@ let () =
     )
     "Print the signature of the corresponding type constructor."
 
+let () =
+  reg_show_prim "show_constructor"
+    (fun env loc id lid ->
+       let desc = Env.lookup_constructor ~loc Env.Positive lid env in
+       if (Ctype.equal env true [desc.cstr_res] [Predef.type_exn]) then
+         raise Not_found;
+       let path =
+         match Ctype.repr desc.cstr_res with
+         | {desc=Tconstr(path, _, _)} -> path
+         | _ -> raise Not_found
+       in
+       let type_decl = Env.find_type path env in
+       [ Sig_type (id, type_decl, Trec_not, Exported) ]
```

Comment:
Your issue is that you are using the constructor `id` . Since you have the type `path`, if the path is just `Pident id`,  you can use this `Ident.t` directly. Otherwise, you need to synthetize a false `Ident.t` name from this path by concatenating all components of the path.

Reply:
ah ok, I confess I thought it might be horribly dangerous to make up idents as it seemed to have side-effects (ie incrementing currentstamp) and I didn't know whether this reached further.  I'll push up some changes, thanks for the help.

Reply:
If you have the correct type declaration in `type_decl`, you should be able to use `type_decl.typ_id` to get the identifier of the type.

Reply:
@gasche it seems that typ_id lives on Typedtree.type_declaration (part of pattern) where-as I am using Env.find_type which returns a Types.type_declaration - which crucially does not have a Ident.t yet.  Not sure if there is an easy way to get from Types to Typedtree, I guess that is what the type checker is doing?

Reply:
Ah, my mistake. Thinking about the issue further:

- If the type path is `Foo.t`, it could be nice to show `type Foo.t = ...` (even though this is not valid OCaml declaration syntax), as it shows where the type is declared (I believe that `#show Foo.constructor;;`: in the general case the type may not be available in the current scope). We have the information at hand (it's the path), but the format of returning a valid signature item constrains us a bit; but we can make up a fake identifier with a dot in the name if we want.
- The code of the `show_prim` primitive already makes up a fake identifier: `let id = Ident.create_persistent s in ...`. This was enough for most need, but not this one where we want to talk about another identifier than the one corresponding to the Longident.t given by the user. So we could reuse the same hack and use `Ident.create_persistent (Path.name path)`.

Reply:
Ah ok, i see that create_persistent doesn't modify the currentstamp so has no side-effects, that seems like a better approach than create_local.